### DISCUSSION TITLE
GitHub command line access question

### DISCUSSION THREAD

**Discussion by brian-math** [Audiveris/audiveris#567] [General]
I'm trying to push a new local git branch "drum-notation" to GitHub Audiveris project from Mac Terminal, but getting an error:

% git push -u origin drum-notation
[I enter my GitHub username and personal-access-token when prompted]
remote: Permission to Audiveris/audiveris.git denied to brian-math.
fatal: unable to access 'https://github.com/Audiveris/audiveris.git/': The requested URL returned error: 403

% git remote -v
origin	https://github.com/Audiveris/audiveris.git (fetch)
origin	https://github.com/Audiveris/audiveris.git (push)

I deleted the GitHub.com entry in Mac Keychain Access, although when I logged out and back in to GitHub.com, a new keychain entry was created automatically by Safari, but this one is suffixed with "(Passwords not saved)" and does not contain either my GitHub username or password.

Suggestions? @maximumspatium ?

**Comment by ecstrema** [comment]
You need to fork the repository, set the origin remote (this should be brian-math/audiveris) to your fork, then push to your fork.

You'll then be able to open a pull request to the upstream (audiveris/audiveris) repo.

**Reply by ecstrema** [reply]
https://twitter.com/TonyJHerbert22/status/1511824744940720138

**Comment by brian-math** [comment]
Thank you @Marr11317 ! That did the trick :-)

**Comment by brian-math** [comment]
[I have moved this conversation out of #33 since it has nothing to do with drum notation.]

@Jojo-Schmitz wrote

> Try `git pull --rebase upstream development` (provided your upstream's name is 'upstream')
> Then a `git push -f`

> No, "origin" is your fork on GitHub. If git remote -v doesn't also show https://github.com/Audiveris/audiveris.git, something is wrong there, and you'd need to add that.
>
>I believe it'd be `git remote add upstream https://github.com/Audiveris/audiveris.git`

After backing up locally the files I changed on my branch, and viewing some log information about my current git branch status, I followed your instructions (making sure I was currently on my drum-notation branch locally):

```
% git remote add upstream https://github.com/Audiveris/audiveris.git
% git remote -v
origin	https://github.com/brian-math/audiveris.git (fetch)
origin	https://github.com/brian-math/audiveris.git (push)
upstream	https://github.com/Audiveris/audiveris.git (fetch)
upstream	https://github.com/Audiveris/audiveris.git (push)

% git branch -vv
  development   1e3432281 [origin/development] Use of closeable within ProxyMusic
 * drum-notation e65885bac [origin/drum-notation] Added diamond noteheads to ShapeSet.getStemTemplateNotes to get stem/beam connections
  master        c09447621 [origin/master] Update Handbook on 5.3 features

% git pull --rebase upstream development
From https://github.com/Audiveris/audiveris
 * branch                development -> FETCH_HEAD
 * [new branch]          development -> upstream/development
Current branch drum-notation is up to date.

% git push -f
Everything up-to-date

```
But nothing much seemed to change. Here are the first few lines of output of `% git log --oneline --decorate --graph --all ` both before and after the above commands:

BEFORE:
\* e65885bac (HEAD -> drum-notation, origin/drum-notation) Added diamond noteheads to ShapeSet.getStemTemplateNotes to get stem/beam connections
\* d2998a546 First steps recognizing diamond (filled & hollow) drum noteheads
\* 1e3432281 (origin/development, development) Use of closeable within ProxyMusic
\* 40492b964 Cross-head shape added to the shapes not crossable by curves
\* 666232209 Allow any non-positive value to manually nullify a scaling item
\* c1a626c76 Bug fixes related to almost empty system

AFTER:
\* e65885bac (HEAD -> drum-notation, origin/drum-notation) Added diamond noteheads to ShapeSet.getStemTemplateNotes to get stem/beam connections
\* d2998a546 First steps recognizing diamond (filled & hollow) drum noteheads
\* 1e3432281 (**upstream/development,** origin/development, development) Use of closeable within ProxyMusic
\* 40492b964 Cross-head shape added to the shapes not crossable by curves
\* 666232209 Allow any non-positive value to manually nullify a scaling item
\* c1a626c76 Bug fixes related to almost empty system

The only thing that has changed is the new pointer upstream/development on the 3rd snapshot. Moreover it seems that drum-notation **is a sub-branch of development** (at least locally) because its snapshot e65885b occurs above (more recently on the timeline, and a direct successor of) the most recent snapshot in development, 1e34322.

However on my personal GitHub account, on the audiveris fork, my drum-notation branch still appears to be a sub-branch of the master branch: when I select my drum-notation branch on the GitHub website, immediately below appears the message:

"This branch is 68 commits ahead, 6 commits behind Audiveris/audiveris:master."

Clicking the Fetch Upstream button produces the message:

"This branch is 6 commits behind the upstream and has conflicts that must be resolved."

And if I start the process of issuing a Pull Request, GitHub defaults to asking to merge with the master branch, not the development branch. So something is still wrong on my GitHub.com fork.

Suggestions?

**Reply by brian-math** [reply]
@hbitteur / Hervé: 

Since nobody has responded to my query from Apr 15 (above), I have tried many things this weekend to try to get GitHub to recognize a drum-notation branch as "This branch is up to date with Audiveris/audiveris:development" rather than "This branch is 66 [or 68] commits ahead, 6 commits behind Audiveris/audiveris:master." Most recently I deleted the old drum-notation branches both on my brian-math GitHub fork of the Audiveris project, and on my local computer, and re-created a brand new drum-notation branch while the local development branch was active. 

In some sense this seems to have been successful: 
% git log --oneline --decorate --graph --all
\* 1e3432281 (HEAD -> drum-notation, upstream/development, origin/drum-notation, origin/development, development) Use of closeable within ProxyMusic
\* 40492b964 Cross-head shape added to the shapes not crossable by curves
...
Notice that the **local** _drum-notation_ and _development_, the **upstream** (Audiveris GitHub account) _development_, and the **origin** (brian-math/audiveris fork) _drum-notation_ and _development_ branches all point to the same commit, namely the latest one made on the development branch. I believe this is the correct and expected behavior.

However, on my brian-math/audiveris project on GitHub.com, selecting the new drum-notation branch still displays the same old message, "This branch is 66 commits ahead, 6 commits behind Audiveris/audiveris:master." When I think about it, though, the latter message does perfectly describe where the development branch is, so it is a true and accurate description of where the drum-notation branch is, too.

Maybe if I re-submit a PR from the new drum-notation branch (which currently has no changes to any files) to Audiveris/audiveris:development, and if you accept the PR with that destination, that might be enough to force GitHub to state that the drum-notation branch is "up to date with Audiveris/audiveris:development" as we would like. I don't think there is much to lose, since as I mentioned I have rolled back all the code changes that I made on the original drum-notation branch (saved as backup files in a separate directory on my local machine, as you suggested).

The only other strategy I can think of is for you to create a new subbranch (maybe call it something different, like "drum-scores") of the development branch on the Audiveris GitHub account, which I could then checkout to my local machine and push to my brian-math fork on GitHub.com. Perhaps such an "official" subbranch would be described by GitHub as "This branch is up to date with Audiveris/audiveris:development", as we want.

Please let me know what you would like me to do...

[Edit: Well, it looks like I can't issue a PR from the brand new drum-notation to Audiveris/audiveris:development. If I try, GitHub says "There isn’t anything to compare. Audiveris:development and brian-math:drum-notation are identical."

So I guess if you want me to issue a PR from the new branch, I'll have to change at least one file. But I could do so in a very innocuous way, e.g. by adding a blank comment at the beginning.]

[Edit#2: I think the 2nd strategy might be better: create a new "official" subbranch off of _development_ in the Audiveris GitHub account for adding capabilities involving percussion notation. That way you can pull changes to the subbranch and keep them there for a while, and only merge them with the _development_ branch when you think enough features have been added and they've been fully tested. Feel free to name the new subbranch _drum-notation_ if you like. I'm happy to delete my existing _drum-notation_ branch and start over with yours.]

**Comment by hbitteur** [comment]
I'm away from home. I plan to address this when I return on Wednesday.

Le dim. 17 avr. 2022 à 22:13, brian-math ***@***.***> a
écrit :

**Comment by maximumspatium** [comment]
@brian-math It looks like I'm late to the party. I'd be happy to help you with Audiveris and Github.

I hope you have a local copy of your work because it has been recently deleted from your remote branch!

I suggest to start with a clean local copy:
```
% cd [your local development directory]
% git clone https://github.com/Audiveris/audiveris.git audiveris-new
% cd audiveris-new
% git checkout development
```
That last command switches your local repository to `development` branch. It's necessary to do that because you should base your work on the recent code that is resided in the `development` branch, not `master`.

After that, you'd need to _**rebase**_ your work onto the freshly cloned `development` branch. Rebase is an action in Git that allows you to rewrite commits from one branch onto another branch. You can do it several ways. The easiest way is probably `git merge`, something like that:

```
% cd audiveris-new
% git checkout development
% git merge [path_to_your_old_audiveris_local_repo] drum-notation
```

If the above suggested `git merge` fails, you could try to move your commits to the new local repository one by one. That's called cherry picking. The corresponding command is `git cherry-pick`.

After rebasing and testing everything, you can proceed with pushing your new local branch to a remote repository of YOURS. Github won't allow you to push changes directly to the remote repository of the Audiveris project. Doing so isn't safe. Github provides you with the ability to submit pull requests instead. This way, Audiveris developers can integrate your work safely by using Github tools that ensure that nothing gets inadvertently broken.

```
% cd audiveris-new
% git remote add origin https://github.com/brian-math/audiveris.git
% git push -f origin development
```

You'll probably need `-f` in the last command in order to overwrite your remote repository.

Finally, you can navigate to your Github repository with your browser and open a pull request there, preferably from `brian-math:audiveris:development` to `audiveris:development`, using the convenient Github GUI.

P.S.: you can move your work to a new branch instead of `development`. If you prefer to work this way, pay attention to create that new branch off the `development` branch so all recent code will be automatically included:

```
% cd audiveris-new
% git checkout development
% git branch drum-notation
% git checkout drum-notation
% [add some commits to drum-notation]
% git push -u origin drum-notation
```

**Reply by brian-math** [reply]
@maximumspatium Thank you very much for your reply and detailed suggestions—I appreciate them.

In an early email to me, Hervé wrote:

• Please have a look at our [Git Workflow](https://github.com/Audiveris/audiveris/wiki/Git-Workflow) wiki article: I'd suggest we create a "drum-notation" branch off of "development" branch for this activity.

So my assumption was that Hervé wanted to have a rather long-lasting (and public?) branch specifically for developing percussion notation capabilities. I was not expecting to continually merge my "drum-notation" branch into my (local and GitHub.com) copy of the "development" branch, and then issue PR's from there to the Audiveris GitHub.com "development" branch. Rather, I thought the "drum-notation" branch would maintain a separate presence, at least on my brian-math GitHub.com account. 

[I just noticed your edit: "P.S...", and that is exactly the work-flow I was envisioning. BTW, the first command under your P.S. should be `% cd audiveris-new`, rather than `% git audiveris-new`, correct?]

I shall wait and see what @hbitteur says, once he returns home tomorrow, about whether "drum-notation" should be a branch off "development" on Audiveris/audiveris or only on brian-math/audiveris.

**Reply by maximumspatium** [reply]
> BTW, the first command under your P.S. should be % cd audiveris-new, rather than % git audiveris-new, correct?

Yes, it should be `cd audiveris-new`, of course.

> whether "drum-notation" should be a branch off "development" on Audiveris/audiveris or only on brian-math/audiveris

It really doesn't matter whether it's going to be a separate branch or not.  The key part is to have your work hosted in a remote repository we can pull code from. `drum-notation` on your Github account is fine.

**Comment by brian-math** [comment]
@maximumspatium:

I followed your instructions and I now have again a drum-notation branch on my brian-math/audiveris GitHub account. GitHub still describes it as "This branch is 67 commits ahead, 6 commits behind Audiveris/audiveris:master." Maybe that's just the way GitHub works (recall that Audiveris/development "is 66 commits ahead, 6 commits behind master", and I made 1 additional commit to drum-notation). Shall I go ahead and issue a pull request to audiveris:development and see what happens?

For the record, here are the local commands I issued, along with the output from a few of them. I can provide the output for every command if that would be helpful.

```
% git clone https://github.com/Audiveris/audiveris.git audiveris-new.
% git remote --v
origin	https://github.com/Audiveris/audiveris.git (fetch)
origin	https://github.com/Audiveris/audiveris.git (push)
% cd audiveris-new
% git checkout development
branch 'development' set up to track 'origin/development'.
Switched to a new branch 'development'
% git branch drum-notation
% git checkout drum-notation
Switched to branch 'drum-notation'
```
At this point I copied several modified .java files to the appropriate subdirectories in audiveris-new

```
% cd ~/Projects/Audiveris/audiveris-new
% git add .
% git log 
commit 1e34322816c52a37a5283ab6ca186898ba90c953 (HEAD -> drum-notation, origin/development, development)
Author: Hervé Bitteur <herve.bitteur@audiveris.org>
Date:   Wed Apr 6 10:34:21 2022 +0200
    Use of closeable within ProxyMusic
% git commit -m "First steps recognizing diamond (filled & hollow) percussion noteheads and getting stem/beam connections"
[drum-notation 8d0f4d204] First steps recognizing diamond (filled & hollow) percussion noteheads and getting stem/beam connections
 8 files changed, 50 insertions(+), 15 deletions(-)
```
Next I tried `% git remote add origin https://github.com/brian-math/audiveris.git` but got an `error: remote origin already exists.` Notice that origin was set to https://github.com/Audiveris/audiveris.git when I first cloned the project into audiveris-new. Instead I did:

```
% git remote rename origin upstream
% git remote add origin https://github.com/brian-math/audiveris.git
% git remote -v
origin	https://github.com/brian-math/audiveris.git (fetch)
origin	https://github.com/brian-math/audiveris.git (push)
upstream	https://github.com/Audiveris/audiveris.git (fetch)
upstream	https://github.com/Audiveris/audiveris.git (push)
% git branch -vv
  development   1e3432281 [upstream/development] Use of closeable within ProxyMusic
* drum-notation 8d0f4d204 First steps recognizing diamond (filled & hollow) percussion noteheads and getting stem/beam connections
  master        c09447621 [upstream/master] Update Handbook on 5.3 features
% git push -u origin drum-notation
% git branch -vv
  development   1e3432281 [upstream/development] Use of closeable within ProxyMusic
* drum-notation 8d0f4d204 [origin/drum-notation] First steps recognizing diamond (filled & hollow) percussion noteheads and getting stem/beam connections
  master        c09447621 [upstream/master] Update Handbook on 5.3 features
```

**Reply by brian-math** [reply]
By the way, if you go to the first commit page for _drum-notation_, https://github.com/brian-math/audiveris/commit/8d0f4d204ce5b6a6ca379e0fcbeea4d09b66acc6, you see that this commit has 1 parent, 1e34322, which happens to be the last commit of the _development_ branch.

I think this proves that _drum-notation_ is a sub-branch of _development_, as desired.

**Comment by hbitteur** [comment]
> So my assumption was that Hervé wanted to have a rather long-lasting (and public?) branch specifically for developing percussion notation capabilities.

Your assumption is true. I had and still have the feeling that a full support of drum notation will not be done in a snap and thus requires a public long-lasting specific branch off of our development branch.

The problem is that this topic is turning into a discussion about git commands I'm far from mastering. I can help you on Audiveris architecture and code, much less on git subtleties.

Here is my proposal:
1. I clone Audiveris from GitHub into a fresh local directory.
2. There, I create a new branch off of development branch. I propose to name it "drum-scores" to avoid any confusion with "your" drum-notation branch.
3. From this drum-scores branch, I try to merge your PR. by a `gh pr checkout 570`
4. And see what I get.

Any advice on this? Until I push back everything to GitHub, there is no risk. if needed I can just discard my local directory.

**Update**: I initially wrote `gh pr merge 570`  by memory, instead of `gh pr checkout 570`, So I'm afraid this will not perform the local merge I had in mind. Please help me on this.

**Comment by hbitteur** [comment]
Bad news.

I tried the proposed actions, and after the `gh pr merge 570` command (yes, merge not checkout), I noticed no modification in my local files. But on GitHub site, files like the Shape.java file have been modified in the development branch.
In other words, it looks like the merge has been performed directly on GitHub and not locally as I thought!
So, while `git` merge command simply merges, `gh` merge command does merge + push?

This is not a disaster, but I would have preferred some local code inspection and testing before this actual merge.
I noticed on [https://github.com/Audiveris/audiveris/pull/570](https://github.com/Audiveris/audiveris/pull/570) page that there is a "Revert" button facing the line on merged commit, but I resisted the temptation to press it...

Can someone (Maxim?) give a hand on this?

**Reply by maximumspatium** [reply]
> Can someone (Maxim?) give a hand on this?

I would discourage the use of those obscure `gh` commands and encourage the use of the command line Git. It works fine on Windows, too. IMHO, Git is the only tool you need to work with Github.

**Reply by maximumspatium** [reply]
> Can someone (Maxim?) give a hand on this?

I just rolled back the `development` branch to your last commit 14 days ago. Don't forget to update your local copies using the following Git commands:

```
git fetch
git checkout development
git pull
```

**Reply by maximumspatium** [reply]
Done. https://github.com/Audiveris/audiveris/tree/drum-scores

**Reply by maximumspatium** [reply]
Don't forget to execute `git fetch` and `git pull` for your local Audiveris repositories!

**Reply by hbitteur** [reply]
Thanks Maxim for cleaning up my mess, one more time...

I never used `git fetch`. I'm used to `git pull --all`, which seems to perform some fetch beforehand
Is there any difference?

**Reply by brian-math** [reply]
Great, Maxim! I see you made a new branch drum-scores and pulled my commit to it. From now on I'll issue PRs from my brian-math/drum-notation to the main Audiveris/drum-scores branch.

**Reply by Jojo-Schmitz** [reply]
`git fetch` doesn't pull ;-)

**Reply by hbitteur** [reply]
@brian-math 
Well... yes, this is the current situation (now that Maxim has corrected my merge!).
So let's use this public drum-scores branch for the on-going work on drum notation.

**Reply by hbitteur** [reply]
@Jojo-Schmitz 
I agree. Reading the docs git pull does fetch + merge
I can't remember the role of --all in the git pull command though

**Reply by maximumspatium** [reply]
> I never used git fetch.

Git fetch updates the internal state of a local Git repository. For example, when a new branch is recently added to the remote repository, its local clones won't notice the change without `git fetch`.

**Reply by maximumspatium** [reply]
> I'll issue PRs from my brian-math/drum-notation to the main Audiveris/drum-scores branch.
@brian-math 👍

**Reply by hbitteur** [reply]
> Git fetch updates the internal state of a local Git repository. For example, when a new branch is recently added to the remote repository, its local clones won't notice the change without `git fetch`.

It seems that in a `git pull --all`, the `--all` option is passed to the underlying fetch, which then fetches all branches. Then the merge is performed only on the current branch.

**Reply by maximumspatium** [reply]
@hbitteur I prefer to use Git commands that perform only ONE action just like the famous Unix philosophy: do one thing and do it well. I usually break down some complex Git action into several simple actions and use the appropriate Git commands. This way, I can greatly minimize the risk of doing something wrong. Plus, each and every Git action can be easily rolled back using a couple of simple commands.

**Reply by brian-math** [reply]
@maximumspatium : I set my local branch drum-notation to track 'upstream drum-scores', but pushed to 'origin drum-notation' (where upstream = https://github.com/Audiveris/audiveris.git and origin = https://github.com/brian-math/audiveris.git). I hope this is correct--it seemed the right way to proceed in case other developers also make modifications to the drum-scores branch. 

If there is a way to *always* pull from upstream and push to origin, without having to specify 'origin' as part of the git push command, that would be handy.

I did get a rather surprising message after the various git fetch and git merge commands:

```
% git fetch upstream drum-scores
From https://github.com/Audiveris/audiveris
 * branch                drum-scores -> FETCH_HEAD
 * [new branch]          drum-scores -> upstream/drum-scores
% git branch -u upstream/drum-scores
branch 'drum-notation' set up to track 'upstream/drum-scores'.
% git fetch
From https://github.com/Audiveris/audiveris
   1e3432281..e527f3c47  development -> upstream/development
 + 9c4bcdd42...e527f3c47 tess4       -> upstream/tess4  (forced update)
% git merge
Merge made by the 'ort' strategy.
% git status
On branch drum-notation
Your branch is ahead of 'upstream/drum-scores' by 2 commits.
  (use "git push" to publish your local commits)
```
I don't know why my drum-notation would be 2 commits ahead of drum-scores, since you just pushed my changes from yesterday to the new drum-scores branch. I hadn't made any new commits since then before I issued the above commands.

**Reply by maximumspatium** [reply]
I suggest to synchronize with the upstream's `development` branch on a regular basis. You can accomplish that using the following sequence of Git commands:
```
% cd audiveris-new
% git checkout drum-notation
% git pull --rebase https://github.com/Audiveris/audiveris.git development
% git push
```
The 3rd command will first pull changes from the remote branch, then apply your local changes on top of them i.e. rebasing your local branch. In the most cases, it should work out of the box but sometimes you may need to resolve conflicts manually.

**Comment by brian-math** [comment]
Could someone please remind me how to get my personal GitHub development branch to match my local one, after rebasing my local branch onto the end of the fetched Audiveris Github development branch? 

(Note: origin = github.com/brian-math/audiveris.git, upstream = github.com/Audiveris/audiveris.git)

After the fetch from upstream/development, I had locally:
```
* 8db04b1ca (upstream/development) Use JVM option 'add-exports' to circumvent JigSaw barriers
| * 01a8dad21 (HEAD -> development, origin/development) First efforts at recognizing circle-x note heads
|/  
* ee68a8c30 New version for simile marks to handle marks for 2 and 4 measures. Symbols for 2 and 4 measures include the underlying barline
* c715082d1 Bug fix when some staff header content (clef, key or time) is removed
```
Then I did
`% git rebase upstream development
`
to get
```
* 6fa613973 (HEAD -> development) First efforts at recognizing circle-x note heads
* 8db04b1ca (upstream/development) Use JVM option 'add-exports' to circumvent JigSaw barriers
| * 01a8dad21 (origin/development) First efforts at recognizing circle-x note heads
|/  
* ee68a8c30 New version for simile marks to handle marks for 2 and 4 measures. Symbols for 2 and 4 measures include the underlying barline
* c715082d1 Bug fix when some staff header content (clef, key or time) is removed
```
This is how I wanted my local development branch to look! Now I would like to make origin/development look the same. I believe I am asking how to remove the commit 01a8 from origin/development (which after all is identical to 6fa6); i.e. roll origin/development back to ee68. Then presumably I can just
`% git push origin development
`
to fast forward that branch by adding commits 8db0 and 6fa6 to the head.

Thanks in advance!

**Reply by hbitteur** [reply]
Short answer: I have no idea how to do this. I expected Git gurus could give a hand.

Here is what came to my mind:
1/ You submit a PR so that I can merge your code into my local development branch (which is in sync with origin/development)
2/ I make a commit, followed by a pull --rebase (certainly leading to conflicts)
3/ I resolve the conflicts then push my local  development to origin
4/ You can then pull origin/development into whatever local folder of yours.

This may not be exactly what you are asking for, but at least we could both move ahead from this point.
